keps/sig-api-machinery/20190206-watch-bookmark.md

```diff
@@ -0,0 +1,184 @@
+```
+  Added    EventType = "ADDED"
+  Modified EventType = "MODIFIED"
+  Deleted  EventType = "DELETED"
+  Error    EventType = "ERROR"
+  Bookmark EventType = "BOOKMARK"
+```
+
+Watch event with type Bookmark will represent information that all the objects
+up to a given resourceVersion has been processed for a given watcher. So even
+if the last event of other types contained object with resourceVersion rv1,
+receiving a bookmark with resourceVersion rv2 means that there aren't
+any interesting objects for that watcher in between.
+
+Given that we don't want to wait for v2 version of watch API with that change,
+an obvious requirement for introducing it is backward compatibility. Currently
+watch Event type looks as following:
+```
+  type Event struct {
+    Type EventType
+    Object runtime.Object
+  }
+```
+
+As a result, we will represent bookmark event by setting Bookmark type and
+Object of appropriate type with just ObjectMeta.ResourceVersion field set.
+
+Unfortunately, such a change would break existing clients, as an example
+[official decoder][]. As a result, we will extend `ListOptions` (which is
+how we pass options to watch) with a boolean field where user can opt-in
+for watch bookmarks:
+```
+  type ListOptions struct {
+    ...
+    AllowBookmarks bool
+  }
+```
+We consciously make it just a boolean flag - this gives kube-apiserver an
+ability to choose when they should be send without setting any expectations
+on user side how frequently and when it would be happening. In particular
+client isn't guaranteed to get any bookmarks.
+Such addition to `ListOptions` is also safe from backward-compatibility
+point of view, old kube-apiserver will simply drop this field if set.
+
+Once the API is extended, we add a support for sending bookmarks to watchcache.
+The exact policy of sending them is to be determined, the ideas include:
+- send a bookmark if there weren't any event send to user in the last X
+seconds
+- if we know we will be closing (e.g. due to timeout) a watch, try to send
+a bookmark immediately before that
+
+Finalizing the decision shouldn't block the initial version of this KEP.
+
+[official decoder]: https://github.com/kubernetes/kubernetes/blob/5d4795e14e02ac29273009d86ba3c5012684d5f4/staging/src/k8s.io/client-go/rest/watch/decoder.go#L57
+
+
+### Risks and Mitigations
+
+Sending "watch bookmarks" may break clients not understanding them.
+As a result, we make them explicitly opt-in.
+
+## Graduation Criteria
+
+Beta:
+- Proved scalability/performance gain. With a simple POC I was able to
+reduce amount of processed "init events" by ~40x. So setting the minimum
+goal on 10x without adding any additional visible overhead.
+- Generated informers make use of this new API.
+
+GA:
+- Enabled by default in Kubelet for watching pods for a release.
+- No complaints about the API for a release
+
+## Implementation History
+
+- RRRR-MM-DD: KEP Summary, Moativation and Proposal merged
+
+## Rejected alternatives
```

Comment:
@wojtek-t Can we ease the load on apiserver by increasing the watch cache size for a specific resource (eg: pods) in the cluster version < 1.15 ?